pkg/logs/input/docker/launcher.go

```diff
@@ -159,13 +173,68 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
 	}
 
 	// keep the tailer in track to stop it later on
-	l.tailers[containerID] = tailer
+	l.addTailer(containerID, tailer)
 }
 
 // stopTailer stops the tailer matching the containerID.
 func (l *Launcher) stopTailer(containerID string) {
 	if tailer, isTailed := l.tailers[containerID]; isTailed {
 		go tailer.Stop()
-		delete(l.tailers, containerID)
+		l.removeTailer(containerID)
+	}
+}
+
+func (l *Launcher) restartTailer(containerID string) {
+	log.Warnf("Tailing will restart for %v", ShortContainerID(containerID))
+	backoffDuration := backoffInitialDuration
+	cumulatedBackoffDuration := 0 * time.Second
+	var tailer *Tailer
+	var source *config.LogSource
+
+	for {
+		if backoffDuration > backoffMaxDuration {
+			log.Warnf("Could not restart container %v. Stop tailing", ShortContainerID(containerID))
+			return
+		}
+
+		oldTailer, exists := l.tailers[containerID]
+		if exists {
+			source = oldTailer.source
+			oldTailer.Stop()
```

Comment:
I'm not sure I understand why we stop the old one and create a new one.
Can you explain why we can't re-use it ?

Reply:
There was an issue with a double close of a channel. I think @ajacquemot already had this issue in the udp tailer. We figured it was easier to create a new tailer and let the older get GCed 

```
2018-09-13 12:57:04 UTC | INFO | (tailer.go:93 in Start) | Start tailing container: da1e7badfbe8
2018-09-13 12:57:05 UTC | WARN | (launcher.go:219 in restartTailer) | Could not start tailer for container da1e7badfbe8: error during connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.25/containers/da1e7badfbe81904fbf40de8aa2629aaf648183e054e6db0c1d1e75e11c3e123/logs?follow=1&since=1536843424.788822639&stderr=1&stdout=1&tail=&timestamps=1: read unix @->/var/run/docker.sock: read: connection reset by peer
2018-09-13 12:57:06 UTC | INFO | (tailer.go:93 in Start) | Start tailing container: da1e7badfbe8
panic: close of closed channel

goroutine 228 [running]:
github.com/DataDog/datadog-agent/pkg/logs/decoder.(*Decoder).Stop(0xc4209eeed0)
	/home/vagrant/go/src/github.com/DataDog/datadog-agent/pkg/logs/decoder/decoder.go:96 +0x2e
github.com/DataDog/datadog-agent/pkg/logs/input/docker.(*Tailer).readForever(0xc420172b80)
	/home/vagrant/go/src/github.com/DataDog/datadog-agent/pkg/logs/input/docker/tailer.go:141 +0x4fb
created by github.com/DataDog/datadog-agent/pkg/logs/input/docker.(*Tailer).tail
	/home/vagrant/go/src/github.com/DataDog/datadog-agent/pkg/logs/input/docker/tailer.go:127 +0x13f
```